### PR TITLE
Improve deployment workflow

### PR DESCRIPTION
**Pull request by paphonb**
## Why did you create this PR
Currently, the deploy workflow executes build for all apps, and after pushing to the registry, it also pulls every known image regardless of what's changed.

## What did you do
Simply replacing `nx run-many` to `nx affected` won't work, because `cugetreg-web` will always be affected from renaming the env files. 
To solve this issue, and to future-proof for when we will move to gitops flow soon, I separated change detection, building, and pushing into its own steps.

The new workflow will be as follow:
- First, we get the list of affected apps, even before renaming env files for frontend
- Setup env files and docker
- Build the images
- Tag the freshly built images and push them
- Pull the new images and call docker compose up

In order to simplify the workflow, the `docker-build` commands now only have one `deploy` configuration that outputs an image with the app name as the temporary tag. The images will then be properly tagged and pushed later in the workflow. This is essentially moving the logic out of the project settings, to allow more customizations in the workflow.

## Demo
- [Run 169](https://github.com/thinc-org/cugetreg-frontend/runs/7997001986?check_suite_focus=true) - `cugetreg-web` affected, only `cugetreg-web` built and deployed
- [Run 170](https://github.com/thinc-org/cugetreg-frontend/runs/7997125932?check_suite_focus=true) - None affected, the whole process skipped (we might want to tweak this behavior later)

## Checklist
- [x] Deploy a demo
- [ ] Check browsers compatibility
- [ ] Wrote coverage tests

### CODE REVIEW

File: tools/affected-apps [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
```
Comment (bombnp) [inline tools/affected-apps:1]: why not have `.sh` extension?

File: tools/tag-push-images [hunk 2]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
```
Comment (bombnp) [inline tools/tag-push-images:1]: why not have `.sh` extension?

File: tools/affected-apps (hunk 1 above)
Comment (bombnp) [reply tools/affected-apps:1]: also should make it `set-affected-apps-env.sh`

**Review by bombnp — CHANGES_REQUESTED** [review]

File: tools/affected-apps (hunk 1 above)
Comment (paphonb) [reply tools/affected-apps:1]: It's just my preference. I think this looks better.

Comment (paphonb) [reply tools/affected-apps:1]: But I do agree about the naming

**Review by saenyakorn — APPROVED** [review]

File: tools/tag-push-images (hunk 2 above)
Comment (paphonb) [reply tools/tag-push-images:1]: Same as above. It's just preference

File: tools/affected-apps [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+# A simple script to output the list of affected apps
+
+if AFFECTED_APPS=$(npx nx affected:apps --base=${AFFECTED_BASE} --plain | tail -1); then
```
Comment (bombnp) [inline tools/affected-apps]: what does plain mean?

File: tools/affected-apps (hunk 1 above)
Comment (bombnp) [reply tools/affected-apps]: it's very confusing about what the file is, without looking at its content

File: tools/affected-apps (hunk 3 above)
Comment (paphonb) [reply tools/affected-apps]: Normally, `nx affected:apps` prints out beautified output, which is hard to parse. `--plain` switches the output format to simple strings separated by spaces.

File: tools/tag-push-images (hunk 2 above)
Comment (bombnp) [inline tools/tag-push-images:1]: also, i wonder if we should move this file to be with the workflows, since it's strictly used in CI?

Comment (paphonb) [reply tools/tag-push-images:1]: Mainly because having it as a separate file means you get syntax highlighting. Plus it removes clutter from the workflow file.

File: tools/affected-apps (hunk 1 above)
Comment (paphonb) [reply tools/affected-apps]: IMO it's pretty clear what it is from just the name

File: tools/tag-push-images (hunk 2 above)
Comment (paphonb) [reply tools/tag-push-images:1]: Mainly because having it as a separate file means you get syntax highlighting. Plus it removes clutter from the workflow file.

File: tools/affected-apps (hunk 1 above)
Comment (bombnp) [reply tools/affected-apps]: @paphonb the problem is i don't know that it's an executable, i thought it's just a text file, nothing else

File: tools/tag-push-images (hunk 2 above)
Comment (bombnp) [reply tools/tag-push-images:1]: i meant keep the file near workflow yaml file

**Review by bombnp — APPROVED** [review]
not really anything else tbh
